### PR TITLE
BUG: Bug in DataFrame construction with nulls and datetimes in a list like

### LINKED ISSUE #15869: "IndexError: tuple index out of range" error with numpy array contain datetimes
#### Code Sample, a copy-pastable example if possible

```python
>>> from datetime import datetime as dt
>>> import numpy as np
>>> from pandas import DataFrame
>>> d=np.array([None, None, None, None, dt.now(), None])
>>> b = DataFrame(d)

```
#### Problem description

The above code works in an old version of pandas (0.7.3) but fails in the current version (0.19.2)

#### Expected Output
No error and dataframe containing the numpy date array
#### Output of ``pd.show_versions()``

<details>
>>> pandas.show_versions()
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 26 Stepping 5, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 20.10.1
Cython: None
numpy: 1.12.1
scipy: None
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None
>>>
</details>

### PR DESCRIPTION
**Pull request by jreback**
closes #15869

### CODE REVIEW

File: pandas/types/cast.py [hunk 1]
```diff
@@ -806,25 +806,20 @@ def _try_timedelta(v):
...
         elif inferred_type in ['timedelta', 'timedelta64']:
             value = _try_timedelta(v)
+        elif inferred_type in ['datetime_or_timedelta']:
 
-        # It's possible to have nulls intermixed within the datetime or
-        # timedelta.  These will in general have an inferred_type of 'mixed',
-        # so have to try both datetime and timedelta.
-
-        # try timedelta first to avoid spurious datetime conversions
-        # e.g. '00:00:01' is a timedelta but technically is also a datetime
```
Comment (chrisaycock) [inline pandas/types/cast.py]: This comment adds clarity to why the order of conversion was chosen.

File: pandas/_libs/src/inference.pyx [hunk 2]
```diff
@@ -439,31 +439,45 @@ def infer_dtype(object value):
...
         Py_ssize_t i, n = len(arr)
         bint seen_timedelta = 0, seen_datetime = 0
+        bint seen_nat = 0, seen_string = 0, seen_null = 0
         object v
 
     for i in range(n):
         v = arr[i]
         if util.is_string_object(v):
-            continue
+            seen_string = 1
```
Comment (chrisaycock) [inline pandas/_libs/src/inference.pyx]: Is `seen_string` supposed to be used somewhere?

**Comment by jreback** [comment]
actually this is non-performant.

The issue is that it ends up looking at all of the strings, when we need to just test a couple to see if they are convertible. Not to mention this fails a couple of tests that have to do with us preserving ``date`` (which is annoying).

File: pandas/_libs/src/inference.pyx [hunk 3]
```diff
@@ -439,31 +439,45 @@ def infer_dtype(object value):
...
         elif util._checknull(v):
-            continue
-        elif is_datetime(v):
-            seen_datetime=1
-        elif is_timedelta(v):
-            seen_timedelta=1
+            # nan or None
+            seen_null = 1
```
Comment (chrisaycock) [inline pandas/_libs/src/inference.pyx]: Similarly, is `seen_null` supposed to be used somewhere?

Comment (jreback) [inline pandas/_libs/src/inference.pyx]: no these were tries.......

**Comment by jreback** [comment]
I think this is fixed. boy user convenience is a PITA.

**Comment by jreback** [comment]
ok I think its fixed now.

**Comment by chrisaycock** [comment]
As long as the test/benchmarks pass, then LGTM. This was a bigger effort than I thought it would be.

**Comment by jreback** [comment]
@chrisaycock yeah a lot of this lower-level inference code is reasonable as we have quite a few integration tests, so refactoring stuff that I wrote years ago is pretty straightforward. But of course there are *lots* of cases and easy to break stuff (which the tests catch).

Down the road we will will be able to simplify things. For example ``datetime.date`` is not a first class type, so its *very* hacky when we infer from it (to maintain some compat). In the future, it *will* be more straightforward to do this (as we will have a valid first class type).

Further I have to think about whether we still want things like:

```
In [5]: Series([pd.NaT, '1 day'])
Out[5]: 
0      NaT
1   1 days
dtype: timedelta64[ns]
```

to work. It is reasonable & not too hard.

*this* is actually a bigger issue:

```
In [6]: Series([pd.NaT])
Out[6]: 
0   NaT
dtype: datetime64[ns]
```

as this is somewhat arbitrary. but this is for another time.

**Pull request closed without merging**